### PR TITLE
Remove include dependency on underlying crypto libraries

### LINKED ISSUE #2284: CHIPCryptoPal.h should not have implementation (OpenSSL/mbedTLS) specific header includes
#### Problem
`CHIPCryptoPal.h` includes headers from underlying implementations (OpenSSL, mbedTLS). This requires the code at upper layers to be aware of which crypto implementation is being used, and requires it to specify include paths (e.g. iOS framework project file has to point to `mbedTLS` header file location).

 #### Proposed Solution
The interface defined in the CHIP header file should be independent from underlying implementations. The interface should be cleaned up, and the included files should be moved in the appropriate implementation file.

### PR DESCRIPTION
**Pull request by pan-apple**
#### Problem
CHIPCryptoPal.h includes headers from underlying implementations (OpenSSL, mbedTLS). This requires the code at upper layers to be aware of which crypto implementation is being used, and requires it to specify include paths (e.g. iOS framework project file has to point to mbedTLS header file location).

 #### Summary of Changes
Move implementation specific code to individual implementation source code. The interface has opaque types for the context, and the size of context is checked via `static_assert` in each implementation file.

fixes #2284

### CODE REVIEW

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 1]
```diff
@@ -276,7 +277,10 @@ CHIP_ERROR Hash_SHA256(const unsigned char * data, const size_t data_length, uns
     return error;
 }
 
-Hash_SHA256_stream::Hash_SHA256_stream(void) {}
+Hash_SHA256_stream::Hash_SHA256_stream(void)
+{
+    static_assert(sizeof(mContext) >= sizeof(SHA256_CTX), "Need more memory to store SHA256 context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:282]: Minor nit; however, if this file has no other C++11 dependencies, recommend using the portable static assertion in nlassert.h.

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 2]
```diff
@@ -812,26 +822,36 @@ void ClearSecretData(uint8_t * buf, uint32_t len)
...
 CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::InitInternal(void)
 {
     CHIP_ERROR error  = CHIP_ERROR_INTERNAL;
     int error_openssl = 0;
 
-    context.curve   = NULL;
-    context.bn_ctx  = NULL;
-    context.md_info = NULL;
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:838]: Minor nit; however, if this file has no other C++11 dependencies, recommend using the portable static assertion in nlassert.h.

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 3]
```diff
@@ -887,10 +910,13 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::Mac(const unsigned char * key, size_t
     int error_openssl        = 0;
     unsigned int mac_out_len = 0;
 
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:914]: See above.

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 4]
```diff
@@ -929,8 +955,11 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::FELoad(const unsigned char * in, size_
     int error_openssl = 0;
     BIGNUM * bn_fe    = (BIGNUM *) fe;
 
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:959]: See above.

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 5]
```diff
@@ -968,7 +997,10 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::FEMul(void * fer, const void * fe1, co
     CHIP_ERROR error  = CHIP_ERROR_INTERNAL;
     int error_openssl = 0;
 
-    error_openssl = BN_mod_mul((BIGNUM *) fer, (BIGNUM *) fe1, (BIGNUM *) fe2, (BIGNUM *) order, context.bn_ctx);
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:1001]: See above.

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 6]
```diff
@@ -991,9 +1026,12 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::PointLoad(const unsigned char * in, si
 
 CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::PointWrite(const void * R, unsigned char * out, size_t out_len)
 {
-    CHIP_ERROR error = CHIP_ERROR_INTERNAL;
+    CHIP_ERROR error          = CHIP_ERROR_INTERNAL;
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:1031]: See above.

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 7]
```diff
@@ -1006,7 +1044,10 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::PointMul(void * R, const void * P1, co
     CHIP_ERROR error  = CHIP_ERROR_INTERNAL;
     int error_openssl = 0;
 
-    error_openssl = EC_POINT_mul(context.curve, (EC_POINT *) R, NULL, (EC_POINT *) P1, (BIGNUM *) fe1, context.bn_ctx);
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:1048]: See above.

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 8]
```diff
@@ -1021,7 +1062,10 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::PointAddMul(void * R, const void * P1,
     int error_openssl  = 0;
     EC_POINT * scratch = NULL;
 
-    scratch = EC_POINT_new(context.curve);
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:1066]: See above.

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 9]
```diff
@@ -1044,7 +1088,10 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::PointInvert(void * R)
     CHIP_ERROR error  = CHIP_ERROR_INTERNAL;
     int error_openssl = 0;
 
-    error_openssl = EC_POINT_invert(context.curve, (EC_POINT *) R, context.bn_ctx);
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:1092]: See above.

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 10]
```diff
@@ -1095,7 +1145,10 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::PointIsValid(void * R)
     CHIP_ERROR error  = CHIP_ERROR_INTERNAL;
     int error_openssl = 0;
 
-    error_openssl = EC_POINT_is_on_curve(context.curve, (EC_POINT *) R, context.bn_ctx);
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:1149]: See above.

File: src/crypto/CHIPCryptoPALmbedTLS.cpp [hunk 11]
```diff
@@ -176,7 +177,10 @@ CHIP_ERROR Hash_SHA256(const unsigned char * data, const size_t data_length, uns
     return error;
 }
 
-Hash_SHA256_stream::Hash_SHA256_stream(void) {}
+Hash_SHA256_stream::Hash_SHA256_stream(void)
+{
+    static_assert(sizeof(mContext) >= sizeof(mbedtls_sha256_context), "Need more memory to store SHA256 context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALmbedTLS.cpp:182]: See above.

File: src/crypto/CHIPCryptoPALmbedTLS.cpp [hunk 12]
```diff
@@ -516,45 +526,66 @@ void ClearSecretData(uint8_t * buf, uint32_t len)
...
-    mbedtls_mpi_init(&context.tempbn);
-    w0     = &context.w0;
-    w1     = &context.w1;
-    xy     = &context.xy;
-    tempbn = &context.tempbn;
-
-    G     = &context.curve.G;
-    order = &context.curve.N;
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALmbedTLS.cpp:553]: See above.

File: src/crypto/CHIPCryptoPALmbedTLS.cpp [hunk 13]
```diff
@@ -567,20 +598,23 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::InitInternal(void)
...
 {
-    mbedtls_ecp_point_free(&context.M);
-    mbedtls_ecp_point_free(&context.N);
-    mbedtls_ecp_point_free(&context.X);
-    mbedtls_ecp_point_free(&context.Y);
-    mbedtls_ecp_point_free(&context.L);
-    mbedtls_ecp_point_free(&context.Z);
-    mbedtls_ecp_point_free(&context.V);
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALmbedTLS.cpp:602]: See above.

File: src/crypto/CHIPCryptoPALmbedTLS.cpp [hunk 14]
```diff
@@ -591,7 +625,11 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::Mac(const unsigned char * key, size_t
 
     mbedtls_md_context_t hmac_ctx;
     mbedtls_md_init(&hmac_ctx);
-    result = mbedtls_md_setup(&hmac_ctx, context.md_info, 1);
+
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALmbedTLS.cpp:630]: See above.

File: src/crypto/CHIPCryptoPALmbedTLS.cpp [hunk 15]
```diff
@@ -707,7 +748,10 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::FEMul(void * fer, const void * fe1, co
 
 CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::PointLoad(const unsigned char * in, size_t in_len, void * R)
 {
-    if (mbedtls_ecp_point_read_binary(&context.curve, (mbedtls_ecp_point *) R, in, in_len) != 0)
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALmbedTLS.cpp:752]: See above.

File: src/crypto/CHIPCryptoPALmbedTLS.cpp [hunk 16]
```diff
@@ -720,8 +764,11 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::PointWrite(const void * R, unsigned ch
     memset(out, 0, out_len);
     size_t mbedtls_out_len = out_len;
 
-    if (mbedtls_ecp_point_write_binary(&context.curve, (const mbedtls_ecp_point *) R, MBEDTLS_ECP_PF_UNCOMPRESSED, &mbedtls_out_len,
-                                       out, out_len) != 0)
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALmbedTLS.cpp:768]: See above.

File: src/crypto/CHIPCryptoPALmbedTLS.cpp [hunk 17]
```diff
@@ -731,7 +778,10 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::PointWrite(const void * R, unsigned ch
 
 CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::PointMul(void * R, const void * P1, const void * fe1)
 {
-    if (mbedtls_ecp_mul(&context.curve, (mbedtls_ecp_point *) R, (const mbedtls_mpi *) fe1, (const mbedtls_ecp_point *) P1,
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALmbedTLS.cpp:782]: See above.

File: src/crypto/CHIPCryptoPALmbedTLS.cpp [hunk 18]
```diff
@@ -743,8 +793,10 @@ CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::PointMul(void * R, const void * P1, co
 CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::PointAddMul(void * R, const void * P1, const void * fe1, const void * P2,
                                                       const void * fe2)
 {
+    Spake2p_Context * context = (Spake2p_Context *) &mSpake2pContext;
+    static_assert(sizeof(mSpake2pContext) >= sizeof(Spake2p_Context), "Need more memory for Spake2p Context");
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALmbedTLS.cpp:797]: See above.

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 19]
```diff
@@ -285,7 +289,9 @@ CHIP_ERROR Hash_SHA256_stream::Begin(void)
     CHIP_ERROR error = CHIP_NO_ERROR;
     int result       = 1;
 
-    result = SHA256_Init(&context);
+    SHA256_CTX * context = (SHA256_CTX *) &mContext;
```
Comment (gerickson) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:292]: Recommend C++-style rather than C-style casts.

**Review by gerickson — CHANGES_REQUESTED** [review]
Generally, I'd recommend:

1) Using C++ rather than C-style casts.
2) Seeing if static_asserts can be reduced by including them only in an initialization context. In addition, they are compile- rather than run-time, so I don't think they provide much value scattered about as they are.
3) Would recommend using the more portable static assertions from nlassert.h if there is nothing else about this module that has explicit C++11 dependencies.

**Comment by pan-apple** [comment]
> Generally, I'd recommend:
> 
> 1. Using C++ rather than C-style casts.
> 2. Seeing if static_asserts can be reduced by including them only in an initialization context. In addition, they are compile- rather than run-time, so I don't think they provide much value scattered about as they are.
> 3. Would recommend using the more portable static assertions from nlassert.h if there is nothing else about this module that has explicit C++11 dependencies.

Thanks @gerickson , all these should be addressed now.

**Review by gerickson — APPROVED** [review]

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 20]
```diff
@@ -812,26 +822,42 @@ void ClearSecretData(uint8_t * buf, uint32_t len)
...
+{
+    EC_GROUP * curve;
+    BN_CTX * bn_ctx;
+    const EVP_MD * md_info;
+} Spake2p_Context;
+
+Spake2p_P256_SHA256_HKDF_HMAC::Spake2p_P256_SHA256_HKDF_HMAC(void) :
+    Spake2p(kP256_FE_Length, kP256_Point_Length, kSHA256_Hash_Length)
+{
+    memset(mSpake2pContext, 0, sizeof(mSpake2pContext));
```
Comment (rwalker-apple) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:835]: I think this is redundant, C++ constructors zero-initialize array member variables

File: src/crypto/CHIPCryptoPALOpenSSL.cpp [hunk 21]
```diff
@@ -812,26 +822,42 @@ void ClearSecretData(uint8_t * buf, uint32_t len)
...
+
 CHIP_ERROR Spake2p_P256_SHA256_HKDF_HMAC::InitInternal(void)
 {
     CHIP_ERROR error  = CHIP_ERROR_INTERNAL;
     int error_openssl = 0;
 
-    context.curve   = NULL;
-    context.bn_ctx  = NULL;
-    context.md_info = NULL;
+    Spake2p_Context * context = reinterpret_cast<Spake2p_Context *>(mSpake2pContext);
```
Comment (rwalker-apple) [inline src/crypto/CHIPCryptoPALOpenSSL.cpp:844]: what if you wrote little static inline functions that did the check, the cast, and returned the appropriate type, making it clearer the requirements for the variable, putting all the hacky stuff together?

File: src/crypto/CHIPCryptoPALmbedTLS.cpp [hunk 22]
```diff
@@ -516,45 +526,72 @@ void ClearSecretData(uint8_t * buf, uint32_t len)
...
+    mbedtls_mpi w0;
+    mbedtls_mpi w1;
+    mbedtls_mpi xy;
+    mbedtls_mpi tempbn;
+} Spake2p_Context;
+
+Spake2p_P256_SHA256_HKDF_HMAC::Spake2p_P256_SHA256_HKDF_HMAC(void) :
+    Spake2p(kP256_FE_Length, kP256_Point_Length, kSHA256_Hash_Length)
+{
+    memset(mSpake2pContext, 0, sizeof(mSpake2pContext));
```
Comment (rwalker-apple) [inline src/crypto/CHIPCryptoPALmbedTLS.cpp:550]: I think memset() to zero is redundant in a constructor

File: src/darwin/Framework/CHIP.xcodeproj/project.pbxproj [hunk 23]
```diff
@@ -454,7 +454,6 @@
 					"CHIP_DEVICE_LAYER_TARGET_LINUX=0",
 					"CHIP_DEVICE_LAYER_TARGET_NRF5=0",
 					"CHIP_DEVICE_LAYER_TARGET_EFR32=0",
-					CHIP_CRYPTO_MBEDTLS,
```
Comment (rwalker-apple) [inline src/darwin/Framework/CHIP.xcodeproj/project.pbxproj:457]: sooo pretty!

**Review by rwalker-apple — APPROVED** [review]
lgtm, little nits

**Review by rwalker-apple — APPROVED** [review]
nice

**Review by andy31415 — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Comment by rwalker-apple** [comment]
@woody-apple needs a button pusher

**Comment by pan-apple** [comment]
@saurabhst, @jelderton, @BroderickCarlin, any comments?

**Review by saurabhst — APPROVED** [review]

**Review by BroderickCarlin — APPROVED** [review]

**Pull request merged**